platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,669 @@
+initializing itself with the correct round. Reconnect therefore is the responsibility of Execution. Consensus does not
+have to consider reconnect at all.
+
+## Modules
+
+### Gossip
+
+The Gossip module is responsible for gossiping messages between neighbors. The actual gossip implementation is not
+described here, except to say that it will be possible to define and implement both event-aware and event-agnostic
+gossip implementations either to a fully connected network or where the set of neighbors is a subset of the whole. This
+document does not dictate whether raw TCP, UDP, HTTP2, gRPC, or other network protocols are used. This will be left to
+the design documents for Gossip.
+
+![Gossip](gossip-module.png)
+
+Gossip is the only part of Consensus that communicates over the network with gossip neighbors. When Gossip is
+initialized, it is supplied a roster. This roster contains the full set of nodes participating in gossip, along with
+their metadata such as RSA signing keys, IP addresses, and so forth. The Gossip module decides which neighbors to gossip
+with (using whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Some gossip algorithms send events in *topological order*. A neighbor may still receive events out of order, because
+different events may arrive from different neighbors at different times. Events received by Gossip are not immediately
+retransmitted to its neighbors. **An honest node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip neighbors and discipline them accordingly. For this reason, events
+received by Gossip are sent to Event Intake, which eventually send valid, ordered events _back_ to Gossip for
+redistribution.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its neighbors for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its neighbors ask it for events, it is able to
+send them. The Gossip module **may** cache all non-expired events, but **must** cache all non-ancient events.
```

Comment:
Is it true the gossip **must** cache events, or it is just more performant to do so? Gossip could theoretically get events from the hashgraph module as needed.

Reply:
> Gossip could theoretically get events from the hashgraph module as needed.

This was the original architecture. The complexities required to maintain a shared data structure was atrocious. From a performance perspective, the old architecture also required a lot of locks and thread synchronization. Although I didn't observe any direct evidence of this causing slowdown, it seemed like a problem waiting to happen.

Reply:
@poulok I think we get a good cleanliness from keeping the modules decoupled in this way. Each module is responsible for its own data structures, and only communicate immutable objects between modules. This is really helpful for understanding the concurrency. But this means instead of a "nexus" of data, we have duplicate data structures. I think the overhead for this is justified.

In this design then, I avoided a "nexus". So then Gossip must cache non-ancient events, or otherwise the gossip break downs entirely and getting to consensus would be impossible in any case where sync was required.

Reply:
I 100% agree that what is stated is the best solution. I just disagree that it is the only solution.